drivers/secure_os/ws_api.cr

```diff
@@ -0,0 +1,167 @@
+require "placeos-driver"
+require "./ws_api_models"
+
+# docs: https://drive.google.com/file/d/1moo9NnFWukSf6fegxaZnSP5ShqSr_A03/view?usp=sharing
+
+class SecureOS::WsApi < PlaceOS::Driver
+  generic_name :SecureOS
+  descriptive_name "SecureOS WebSocket API"
+
+  uri_base "ws://secureos.server:8888/"
+  default_settings({
+    shared_host:   true,
+    rest_api_host: "http://172.16.1.120:8888",
+    basic_auth:    {
+      username: "srvc_acct",
+      password: "password!",
+    },
+  })
+
+  @rest_api_host : String = ""
+  @camera_list : Array(Camera)?
+
+  getter! basic_auth : NamedTuple(username: String, password: String)
+
+  def on_load
+    on_update
+  end
+
+  def on_update
+    @rest_api_host = setting String, :rest_api_host
+    @basic_auth = setting NamedTuple(username: String, password: String), :basic_auth
+  end
+
+  def connected
+    host = setting?(Bool, :shared_host) ? config.uri.not_nil! : @rest_api_host
+    client = HTTP::Client.new URI.parse(host)
+    client.basic_auth **basic_auth
+    response = client.get "#{@rest_api_host}/api/v1/ws_auth"
+    if response.success?
+      json_body = JSON.parse response.body
+      token = json_body["data"]["token"].as_s
+
+      send({type: :auth, token: token}.to_json, wait: false)
+    else
+      raise "Authentication failed"
+    end
+
+    camera_list
+    subscribe_states
+    subscribe_events
+
+    schedule.every(30.seconds) { send({type: :get_server_time}.to_json, name: :server_time) }
+    schedule.every(60.seconds) { camera_list }
+    schedule.every(60.seconds) { subscribe_states }
+    schedule.every(60.seconds) { subscribe_events }
```

Comment:
I would have a single block here:

```suggestion
    schedule.every(5.minutes, immediate: true) do
      camera_list
      subscribe_states
      subscribe_events
    end
```

the `immediate: true` means it'll run straight away